pbm/restore/physical.go

```diff
@@ -0,0 +1,885 @@
+package restore
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	slog "log"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"golang.org/x/mod/semver"
+
+	"github.com/percona/percona-backup-mongodb/pbm"
+	"github.com/percona/percona-backup-mongodb/pbm/log"
+	"github.com/percona/percona-backup-mongodb/pbm/storage"
+)
+
+const (
+	defaultRSdbpath   = "/data/db"
+	defaultCSRSdbpath = "/data/configdb"
+
+	mongofslock = "mongod.lock"
+
+	defaultPort = 27017
+)
+
+type PhysRestore struct {
+	cn     *pbm.PBM
+	node   *pbm.Node
+	dbpath string
+	port   int
+	// an ephemeral port to restart mongod on during the restore
+	tmpPort string
+	rsConf  *pbm.RSConfig // original replset config
+
+	name     string
+	opid     string
+	nodeInfo *pbm.NodeInfo
+	stg      storage.Storage
+	bcp      *pbm.BackupMeta
+	// Shards to participate in restore. Num of shards in bcp could
+	// be less than in the cluster and this is ok.
+	//
+	// Only the restore leader would have this info.
+	shards []pbm.Shard
+
+	stgpath string
+
+	stopHB chan struct{}
+
+	log *log.Event
+}
+
+func NewPhysical(cn *pbm.PBM, node *pbm.Node, inf *pbm.NodeInfo) (*PhysRestore, error) {
+	opts, err := node.GetOpts()
+	if err != nil {
+		return nil, errors.Wrap(err, "get mongo options")
+	}
+	p := opts.Storage.DBpath
+	if p == "" {
+		switch inf.ReplsetRole() {
+		case pbm.RoleConfigSrv:
+			p = defaultCSRSdbpath
+		default:
+			p = defaultRSdbpath
+		}
+	}
+	if opts.Net.Port == 0 {
+		opts.Net.Port = defaultPort
+	}
+
+	rcf, err := node.GetRSconf()
+	if err != nil {
+		return nil, errors.Wrap(err, "get replset config")
+	}
+
+	if inf.SetName == "" {
+		return nil, errors.New("undefined replica set")
+	}
+
+	return &PhysRestore{
+		cn:       cn,
+		node:     node,
+		dbpath:   p,
+		rsConf:   rcf,
+		nodeInfo: inf,
+		port:     opts.Net.Port,
+		tmpPort:  strconv.Itoa(opts.Net.Port + 1111),
```

Comment:
1. I'd try to check if the port is available. if closed try the next one
2. it's possible that `mongod` runs on **65000**

Reply:
Agree. Will do for the GA